File: leetcode-js/052-n-queens-ii.js
/**
 * @param {number} n
 * @return {number}
 */
var totalNQueens = function(n) {
    return solve(0, 0, n, new Set(), new Set(), new Set());
};

function solve(row, count, n, occupiedCols, occupiedDiag1s, occupiedDiag2s) {
    for (let col = 0; col < n; col += 1) {
        if (occupiedCols.has(col)) {
            continue;
        }
        const diag1 = row - col;
        if (occupiedDiag1s.has(diag1)) {
            continue;
        }
        const diag2 = row + col;
        if (occupiedDiag2s.has(diag2)) {
            continue;
        }
        if (row === n - 1) {
            count += 1;
        } else {
            occupiedCols.add(col);
            occupiedDiag1s.add(diag1);
            occupiedDiag2s.add(diag2);
            count = solve(row + 1, count, n, occupiedCols, occupiedDiag1s, occupiedDiag2s);
            occupiedCols.delete(col);
            occupiedDiag1s.delete(diag1);
            occupiedDiag2s.delete(diag2);
        }
    }

    return count;
}

File: leetcode-java/052-n-queens-ii.java
public class Solution {
    public int totalNQueens(int n) {
        return helper(n, 0, 0, new HashSet<Integer>(), new HashSet<Integer>(), new HashSet<Integer>());
    }

    private int helper(int n, int row, int count, Set<Integer> cols, Set<Integer> diag1s, Set<Integer> diag2s) {
        for (int col = 0; col < n; col++) {
            if (cols.contains(col)) {
                continue;
            }

            int diag1 = row - col;
            if (diag1s.contains(diag1)) {
                continue;
            }

            int diag2 = row + col;
            if (diag2s.contains(diag2)) {
                continue;
            }

            if (row == n - 1) {
                count += 1;
            } else {
                cols.add(col);
                diag1s.add(diag1);
                diag2s.add(diag2);

                count = helper(n, row + 1, count, cols, diag1s, diag2s);

                diag2s.remove(diag2);
                diag1s.remove(diag1);
                cols.remove(col);
            }
        }

        return count;
    }
}
